ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -680,9 +685,98 @@ public FindResult findEntriesSince(String partitionPath, FindToken findToken, lo
+      } else if (metadata.containsKey(CloudBlobMetadata.FIELD_EXPIRATION_TIME)) {
+        long expirationTime = Long.parseLong(metadata.get(CloudBlobMetadata.FIELD_EXPIRATION_TIME));
+        eraseBlob = (expirationTime + gracePeriod) < now;
+        eraseReason = String.format("%s: (%s + %s) < %s", CloudBlobMetadata.FIELD_EXPIRATION_TIME, expirationTime, gracePeriod, now);
+      }
+
+      if (eraseBlob) {
+        if (cloudConfig.cloudCompactionDryRunEnabled) {
+          logger.trace("[DRY-RUN] Can erase blob {} from Azure blob storage because {}", blobItem.getName(), eraseReason);
+        } else {
+          Timer.Context storageTimer = azureMetrics.blobCompactionLatency.time();
+          logger.trace("Erasing blob {} from Azure blob storage because {}", blobItem.getName(), eraseReason);
+          blobContainerClient.getBlobClient(blobItem.getName()).delete();
+          storageTimer.stop();
+          vcrMetrics.blobCompactionRate.mark();
+        }
+        numBlobsPurged += 1;
+      } else {
+        logger.trace("Cannot erase blob {} from Azure blob storage because condition not met: {}", blobItem.getName(), eraseReason);
+      }
+
+    }
+    return numBlobsPurged;
+  }
+
+  /**
+   * Compacts a partition
+   * @param partitionPath the path of the partitions to compact.
+   * @return the number of blobs erased from cloud
+   * @throws CloudStorageException
+   */
   @Override
   public int compactPartition(String partitionPath) throws CloudStorageException {
-    return 0;
+    /*
+      BlobContainerStrategy must be PARTITION, otherwise compaction will not work because it cannot find the container in ABS.
+      For BlobContainerStrategy to be CONTAINER, we need a blobId to extract accountId and containerId and we don't have that here.
+     */
+    if (AzureBlobLayoutStrategy.BlobContainerStrategy.get(azureCloudConfig.azureBlobContainerStrategy) !=
+        AzureBlobLayoutStrategy.BlobContainerStrategy.PARTITION) {
+      logger.info("Unable to compact because BlobContainerStrategy is {} when it must be {}", azureCloudConfig.azureBlobContainerStrategy,
+          AzureBlobLayoutStrategy.BlobContainerStrategy.PARTITION);
+      return 0;
+    }
+    String containerName = azureBlobLayoutStrategy.getClusterAwareAzureContainerName(partitionPath);
+    BlobContainerClient blobContainerClient = createOrGetBlobStore(containerName);
+    ListBlobsOptions listBlobsOptions = new ListBlobsOptions().setDetails(new BlobListDetails()
+        .setRetrieveMetadata(true)).setMaxResultsPerPage(azureCloudConfig.azureBlobStorageMaxResultsPerPage);
+    String continuationToken = null;
+    int numBlobsPurged = 0, totalNumBlobs = 0;
+    Timer.Context storageTimer = azureMetrics.partitionCompactionLatency.time();
+    try {
+      logger.info("Initiating compaction of partition {} in Azure blob storage", containerName);
```

Comment:
The compaction can take quite some time.  Background operation shouldn't block shutdown.
Do you plan to quit the compaction when shutdown comes in?

Reply:
good point ! fixed.